File: frontend/src/lib/permissions.ts
export type UserRole = 'Admin' | 'Director' | 'HR Officer' | 'Finance Officer' | 'Staff';

export type Permission = 
  | 'hr:manage'
  | 'hr:attendance:self'
  | 'hr:attendance:manage'
  | 'payroll:generate'
  | 'payroll:approve'
  | 'finance:read'
  | 'finance:write'
  | 'inventory:read'
  | 'inventory:issue'
  | 'programs:read'
  | 'programs:manage'
  | 'admin:users'
  | 'audit:read';

// NOTE: Must stay in sync with backend/app/core/permissions.py
export const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  Admin: [
    'hr:manage', 'hr:attendance:self', 'hr:attendance:manage',
    'payroll:generate', 'payroll:approve', 'finance:read',
    'finance:write', 'inventory:read', 'inventory:issue',
    'programs:read', 'programs:manage', 'admin:users', 'audit:read'
  ],
  Director: [
    'finance:read', 'payroll:approve', 'programs:read',
    'inventory:read', 'audit:read'
  ],
  'HR Officer': [
    'hr:manage', 'hr:attendance:manage', 'hr:attendance:self',
    'payroll:generate'
  ],
  'Finance Officer': [
    'finance:read', 'finance:write', 'payroll:generate',
    'inventory:read'
  ],
  Staff: [
    'hr:attendance:self', 'programs:read', 'inventory:issue'
  ]
};

export function hasPermission(role: string, permission: Permission): boolean {
  if (!role || !(role in ROLE_PERMISSIONS)) return false;
  return ROLE_PERMISSIONS[role as UserRole].includes(permission);
}

File: backend/app/core/permissions.py
from enum import Enum

class UserRole(str, Enum):
    ADMIN = 'Admin'
    DIRECTOR = 'Director'
    HR_OFFICER = 'HR Officer'
    FINANCE_OFFICER = 'Finance Officer'
    STAFF = 'Staff'

class Permission(str, Enum):
    HR_MANAGE = 'hr:manage'
    HR_ATTENDANCE_SELF = 'hr:attendance:self'
    HR_ATTENDANCE_MANAGE = 'hr:attendance:manage'
    PAYROLL_GENERATE = 'payroll:generate'
    PAYROLL_APPROVE = 'payroll:approve'
    FINANCE_READ = 'finance:read'
    FINANCE_WRITE = 'finance:write'
    INVENTORY_READ = 'inventory:read'
    INVENTORY_ISSUE = 'inventory:issue'
    PROGRAMS_READ = 'programs:read'
    PROGRAMS_MANAGE = 'programs:manage'
    ADMIN_USERS = 'admin:users'
    AUDIT_READ = 'audit:read'

# NOTE: Must stay in sync with frontend/src/lib/permissions.ts
ROLE_PERMISSIONS = {
    UserRole.ADMIN.value: [
        Permission.HR_MANAGE.value, Permission.HR_ATTENDANCE_SELF.value, Permission.HR_ATTENDANCE_MANAGE.value,
        Permission.PAYROLL_GENERATE.value, Permission.PAYROLL_APPROVE.value, Permission.FINANCE_READ.value,
        Permission.FINANCE_WRITE.value, Permission.INVENTORY_READ.value, Permission.INVENTORY_ISSUE.value,
        Permission.PROGRAMS_READ.value, Permission.PROGRAMS_MANAGE.value, Permission.ADMIN_USERS.value, Permission.AUDIT_READ.value
    ],
    UserRole.DIRECTOR.value: [
        Permission.FINANCE_READ.value, Permission.PAYROLL_APPROVE.value, Permission.PROGRAMS_READ.value,
        Permission.INVENTORY_READ.value, Permission.AUDIT_READ.value
    ],
    UserRole.HR_OFFICER.value: [
        Permission.HR_MANAGE.value, Permission.HR_ATTENDANCE_MANAGE.value, Permission.HR_ATTENDANCE_SELF.value,
        Permission.PAYROLL_GENERATE.value
    ],
    UserRole.FINANCE_OFFICER.value: [
        Permission.FINANCE_READ.value, Permission.FINANCE_WRITE.value, Permission.PAYROLL_GENERATE.value,
        Permission.INVENTORY_READ.value
    ],
    UserRole.STAFF.value: [
        Permission.HR_ATTENDANCE_SELF.value, Permission.PROGRAMS_READ.value, Permission.INVENTORY_ISSUE.value
    ]
}

def has_permission(role: str, permission: str) -> bool:
    role_perms = ROLE_PERMISSIONS.get(role, [])
    return permission in role_perms
